send-transaction-service/src/send_transaction_service.rs

```diff
@@ -109,100 +147,123 @@ impl SendTransactionService {
-                        } else {
+                        }
+                        if !new_txn {
                             inc_new_counter_info!("send_transaction_service-recv-duplicate", 1);
                         }
                     }
                 }
 
-                if last_status_check.elapsed().as_millis() as u64 >= config.retry_rate_ms {
-                    if !transactions.is_empty() {
-                        datapoint_info!(
-                            "send_transaction_service-queue-size",
-                            ("len", transactions.len(), i64)
-                        );
-                        let (root_bank, working_bank) = {
-                            let bank_forks = bank_forks.read().unwrap();
-                            (
-                                bank_forks.root_bank().clone(),
-                                bank_forks.working_bank().clone(),
-                            )
-                        };
-
-                        let _result = Self::process_transactions(
-                            &working_bank,
-                            &root_bank,
-                            &tpu_address,
-                            &mut transactions,
-                            &leader_info,
-                            &config,
-                        );
+                if !transactions.is_empty()
+                    && (last_batch_sent.elapsed().as_millis() as u64 >= config.batch_send_rate_ms
+                        || transactions.len() >= config.batch_size)
+                {
+                    inc_new_counter_info!(
+                        "send_transaction_service-batch-size",
+                        transactions.len()
+                    );
+                    let _result = Self::send_transactions_in_batch(
+                        &tpu_address,
+                        &mut transactions,
+                        &leader_info,
+                        &config,
+                    );
+                    let last_sent_time = Instant::now();
+                    let mut retry_transactions = retry_transactions.lock().unwrap();
+                    for (signature, mut transaction_info) in transactions.drain() {
+                        let retry_len = retry_transactions.len();
+                        let entry = retry_transactions.entry(signature);
+                        if let Entry::Vacant(_) = entry {
+                            if retry_len >= MAX_TRANSACTION_QUEUE_SIZE {
+                                datapoint_warn!("send_transaction_service-queue-overflow");
+                                break;
+                            } else {
+                                transaction_info.last_sent_time = Some(last_sent_time);
+                                entry.or_insert(transaction_info);
+                            }
+                        }
                     }
-                    last_status_check = Instant::now();
+                    drop(retry_transactions);
```

Comment:
Does this need an explicit drop, or can we simply have the lock enclosed in its own scope?

Reply:
We do not need the explicit drop -- as it will be dropped when going out of scope.  It is done so to trigger a little earlier unlock to reduce the duration the lock is held.

Reply:
Based on Lijun's comments, I agree that switching to using a channel is not a must-have right now, but I do think we can make this cleaner by having the lock in its own scope rather than using an explicit drop:
{
                    let mut retry_transactions = retry_transactions.lock().unwrap();
                    for (signature, mut transaction_info) in transactions.drain() {
                        let retry_len = retry_transactions.len();
                        let entry = retry_transactions.entry(signature);
                        if let Entry::Vacant(_) = entry {
                            if retry_len >= MAX_TRANSACTION_QUEUE_SIZE {
                                datapoint_warn!("send_transaction_service-queue-overflow");
                                break;
                            } else {
                                transaction_info.last_sent_time = Some(last_sent_time);
                                entry.or_insert(transaction_info);
                            }
                        }
                    }

}